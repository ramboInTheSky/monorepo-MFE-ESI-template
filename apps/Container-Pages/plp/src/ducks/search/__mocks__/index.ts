import {mockState} from "../../../../__mocks__/mockStore"

import {selectHasNextPage, selectHasPreviousPage} from ".."

import {selectFetchTriggerOffset, selectSubsequentItemsPerPage} from "../utils"

jest.mock("../utils", () => ({
    selectFetchTriggerOffset: jest.fn(),
    selectSubsequentItemsPerPage: jest.fn(),
}))
jest.mock("..", () => ({
    fixFacets: jest.fn(() => ({type: "FIX_FACETS"})),
    unfixFacets: jest.fn(() => ({type: "UNFIX_FACETS"})),
    selectItems: jest.fn(),
    selectHasNextPage: jest.fn(),
    selectHasPreviousPage: jest.fn(),
    selectCurrentEndPage: jest.fn(),
    selectCurrentStartPage: jest.fn(),
    selectTotalItemsPerPage: jest.fn(),
    fetchNextPageItemsThunk: jest.fn(() => ({type: "foo"})),
    fetchPreviousPageItemsThunk: jest.fn(() => ({type: "foo"})),
}))

export function mockSelectHasNextPage(result) {
    ;(selectHasNextPage as jest.Mock).mockReturnValue(result)
}

export function mockSelectHasPreviousPage(result) {
    ;(selectHasPreviousPage as jest.Mock).mockReturnValue(result)
}

export function mockSelectFetchTriggerOffset(result) {
    ;(selectFetchTriggerOffset as jest.Mock).mockReturnValue(result)
}
export function mockSelectSubsequentItemsPerPage(result) {
    ;(selectSubsequentItemsPerPage as jest.Mock).mockReturnValue(result)
}

export function createMockStateForSearch(searchState: Record<any, any> = {}) {
    return {
        ...mockState,
        search: {
            ...mockState.search,
            ...searchState,
        },
    }
}
