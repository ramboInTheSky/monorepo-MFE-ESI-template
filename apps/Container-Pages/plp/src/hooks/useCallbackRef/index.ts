import {useCallback, useState} from "react"

export function useCallbackRef() {
    const [value, setValue] = useState<any>(null)
    const ref = useCallback((node: HTMLElement) => {
        if (node !== null) setValue(node)
    }, [])
    return [value, ref]
}
