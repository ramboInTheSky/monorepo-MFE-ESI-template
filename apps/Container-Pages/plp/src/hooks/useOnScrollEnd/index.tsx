import {useEffect} from "react"

export function useOnScrollEnd(callback: Function, deps: any[] = []) {
    useEffect(() => {
        let timeout
        function handleScroll() {
            clearTimeout(timeout)
            timeout = setTimeout(() => {
                callback()
            }, 66)
        }
        document.addEventListener("scroll", handleScroll)
        return () => document.removeEventListener("scroll", handleScroll)
    }, [...deps]) // eslint-disable-line react-hooks/exhaustive-deps
}
