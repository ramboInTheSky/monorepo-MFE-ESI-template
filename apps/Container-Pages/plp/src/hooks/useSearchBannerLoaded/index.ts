import {useState, useEffect} from "react"
import {
    handleSeoPillsImpression,
    handleVisualPillsImpression,
    setupSeoPillsEventListeners,
    setupVisualPillsEventListeners,
} from "../../utils/searchBanner"

export const useSearchBannerLoaded = (searchBannerHtml: string | null) => {
    const setMobileSearchBannerText = () => {
        const searchBannerText = document.getElementById("search-banner-read-more-content")
        const mobileSearchBannerText = document.getElementById("search-banner-mobile-read-more-content")
        if (searchBannerText && mobileSearchBannerText) {
            mobileSearchBannerText.innerHTML = searchBannerText.innerHTML
        }
    }

    const [searchBannerTextLoaded, setSearchBannerTextLoaded] = useState(false)

    useEffect(() => {
        const hasSearchBannerTextLoaded = () => {
            const searchBannerText = document.getElementById("search-banner-read-more-content")

            if (searchBannerText) {
                setSearchBannerTextLoaded(true)
                setMobileSearchBannerText()
            }

            handleVisualPillsImpression(document)
            handleSeoPillsImpression(document)

            setupSeoPillsEventListeners(document)
            setupVisualPillsEventListeners(document)
        }
        hasSearchBannerTextLoaded()
    }, [searchBannerHtml])

    return searchBannerTextLoaded
}
