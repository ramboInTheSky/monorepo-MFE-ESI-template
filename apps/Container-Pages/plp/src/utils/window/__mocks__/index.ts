import {
    getWindow,
    getDocumentScrollTop,
    removeDocumentScrollListener,
    addDocumentScrollListener,
    isSomewhatInViewport,
    isAboveViewport,
    isBelowViewport,
    IS_BROWSER,
} from ".."

jest.mock("..", () => ({
    getWindow: jest.fn(() => ({})),
    IS_BROWSER: jest.fn(() => true),
    isAboveViewport: jest.fn(),
    isBelowViewport: jest.fn(),
    isSomewhatInViewport: jest.fn(),
    getDocumentScrollTop: jest.fn(),
    addDocumentScrollListener: jest.fn(),
    removeDocumentScrollListener: jest.fn(),
}))

function mockListener(fn: jest.Mock) {
    const api = {
        listener: () => null,
    }
    fn.mockImplementation(listener => {
        api.listener = listener
    })
    return api
}

export function mockGetWindow(value: any = window) {
    ;(getWindow as jest.Mock).mockReturnValue(value)
}

export function mockGetDocumentScrollTop(value: number) {
    ;(getDocumentScrollTop as jest.Mock).mockReturnValue(value)
}

export function mockRemoveDocumentScrollListener() {
    return mockListener(removeDocumentScrollListener as jest.Mock)
}

export function mockAddDocumentScrollListener() {
    return mockListener(addDocumentScrollListener as jest.Mock)
}

export function mockIsSomewhatInViewport(result: boolean) {
    ;(isSomewhatInViewport as jest.Mock).mockReturnValue(result)
}

export function mockIsAboveViewport(result: boolean) {
    ;(isAboveViewport as jest.Mock).mockReturnValue(result)
}

export function mockIsBelowViewport(result: boolean) {
    ;(isBelowViewport as jest.Mock).mockReturnValue(result)
}

export function mockIsBrowser(result: boolean) {
    ;(IS_BROWSER as jest.Mock).mockReturnValue(result)
}
