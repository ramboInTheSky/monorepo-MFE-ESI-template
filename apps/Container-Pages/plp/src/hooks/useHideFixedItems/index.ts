import {useEffect, useState} from "react"
import {useSelector} from "react-redux"

import {useHasReachedFooter} from "./useHasReachedFooter"
import {selectIsFetchingNextPage} from "../../ducks/search"

export function useHideFixedItems() {
    const [hideFixedItems, setHideFixedItems] = useState(false)
    const {hasReachedFooter, calculateHasReachedFooter} = useHasReachedFooter()
    const isFetchingNextPage = useSelector(selectIsFetchingNextPage)

    useEffect(() => {
        if (isFetchingNextPage || !hasReachedFooter) {
            setHideFixedItems(false)
        } else if (!isFetchingNextPage && hasReachedFooter) {
            const isCloseToFooter = calculateHasReachedFooter(true)
            setHideFixedItems(!!isCloseToFooter && hasReachedFooter)
        }
    }, [isFetchingNextPage, hasReachedFooter, calculateHasReachedFooter])

    return {
        hideFixedItems,
    }
}
