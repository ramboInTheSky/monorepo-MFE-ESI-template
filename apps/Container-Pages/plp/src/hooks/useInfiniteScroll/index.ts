import {useCallback, useEffect, useMemo, useRef} from "react"
import {addDocumentScrollListener, removeDocumentScrollListener} from "../../utils/window"
import {useInfiniteScrollHelpers} from "./useInfiniteScrollHelpers"
import {getFetchTriggerElement} from "./utils"
import {DOWN, UP} from "./types"

export interface InfiniteScrollOptions {
    endPage: number
    startPage: number
    totalItems: number
    hasNextPage: boolean
    itemsPerPage: number
    hasPreviousPage: boolean
    isFetchingNextPage: boolean
    isFetchingPreviousPage: boolean
    isFetchingPageItems: boolean
    fetchTriggerOffset?: number
    fetchNextPage: () => Promise<any>
    fetchPreviousPage: () => Promise<any>
    gridContainerElement: HTMLElement | null
    addScrollEventListener?: typeof addDocumentScrollListener
    removeScrollEventListener?: typeof removeDocumentScrollListener
    onBeforeFetchingPreviousPage?: () => void
    onAfterRestoringUpScrollPosition?: () => void
}

export interface InfiniteScrollHookResult {
    handleScroll: () => Promise<void>
}

export function useInfiniteScroll({
    endPage,
    startPage,
    totalItems,
    hasNextPage,
    itemsPerPage,
    hasPreviousPage,
    fetchNextPage,
    fetchPreviousPage,
    isFetchingNextPage,
    isFetchingPreviousPage,
    isFetchingPageItems,
    gridContainerElement,
    fetchTriggerOffset = 8,
    onBeforeFetchingPreviousPage = noop,
    onAfterRestoringUpScrollPosition = noop,
}: InfiniteScrollOptions): InfiniteScrollHookResult {
    const stateApi = useInfiniteScrollState()
    const helpers = useInfiniteScrollHelpers({
        startPage,
        totalItems,
        itemsPerPage,
        gridContainerElement,
    })

    if (isFetchingPageItems) {
        // we need to remove the fetched page first when filtering the page otherwise it would not trigger the next page
        stateApi.clearFetchedPage()
    }

    const handleScroll = useCallback(async () => {
        if (!gridContainerElement) return

        const {
            hasFetchedPage,
            isFetchingPage,
            flagAsFetchingPage,
            flagAsHavingFetchedPage,
            flagAsNoLongerFetchingPage,
        } = stateApi

        const {
            getScrollDirection,
            waitForPageToRerender,
            isTimeToFetchNextPage,
            isTimeToFetchPreviousPage,
            restoreUpScrollPosition,
        } = helpers

        const scrollDirection = getScrollDirection()
        const nextPage = endPage + 1
        const previousPage = startPage - 1
        const cueToTriggerFetch = getFetchTriggerElement(gridContainerElement, fetchTriggerOffset, scrollDirection)

        if (
            hasNextPage &&
            scrollDirection === DOWN &&
            isTimeToFetchNextPage(cueToTriggerFetch) &&
            !hasFetchedPage(nextPage) &&
            !isFetchingPage(nextPage) &&
            !isFetchingNextPage
        ) {
            flagAsFetchingPage(nextPage)
            const attempt = await attemptTo(fetchNextPage)
            flagAsNoLongerFetchingPage(nextPage)
            if (!attempt.failed) flagAsHavingFetchedPage(nextPage)
        }

        if (
            hasPreviousPage &&
            scrollDirection === UP &&
            isTimeToFetchPreviousPage(cueToTriggerFetch) &&
            !hasFetchedPage(previousPage) &&
            !isFetchingPage(previousPage) &&
            !isFetchingPreviousPage
        ) {
            onBeforeFetchingPreviousPage()
            flagAsFetchingPage(previousPage)
            const attempt = await attemptTo(fetchPreviousPage)
            flagAsNoLongerFetchingPage(previousPage)
            if (!attempt.failed) {
                // By the time it gets to this point, the component
                // would not have re-rendered with the newly fetched
                // items because redux would not yet have triggered
                // the update. So we need to explicitly wait for the
                // the rerender to occur before we reposition the
                // scroll bar, in order to allow the user to continue
                // to scroll up to reveal the newly loaded items
                await waitForPageToRerender()
                restoreUpScrollPosition()
                onAfterRestoringUpScrollPosition()
                flagAsHavingFetchedPage(previousPage)
            }
        }
    }, [
        endPage,
        startPage,
        hasNextPage,
        hasPreviousPage,
        fetchNextPage,
        fetchPreviousPage,
        isFetchingNextPage,
        isFetchingPreviousPage,
        onBeforeFetchingPreviousPage,
        onAfterRestoringUpScrollPosition,
        gridContainerElement,
        fetchTriggerOffset,
        stateApi,
        helpers,
    ])

    useEffect(() => {
        addDocumentScrollListener(handleScroll)
        return () => removeDocumentScrollListener(handleScroll)
    }, [handleScroll])

    return {handleScroll}
}

function useInfiniteScrollState() {
    const pagesFetched = useRef<number[]>([])
    const pagesBeingFetched = useRef<number[]>([])

    return useMemo(
        () => ({
            hasFetchedPage: page => {
                return pagesFetched.current.includes(page)
            },
            isFetchingPage: (page: number) => {
                return pagesBeingFetched.current.includes(page)
            },
            flagAsFetchingPage: (page: number) => {
                pagesBeingFetched.current = [...pagesBeingFetched.current, page]
            },
            flagAsHavingFetchedPage: (page: number) => {
                pagesFetched.current = [...pagesFetched.current, page]
            },
            flagAsNoLongerFetchingPage: (page: number) => {
                const newPages = [...pagesBeingFetched.current]
                const indexToRemove = newPages.indexOf(page)
                newPages.splice(indexToRemove, 1)
                pagesBeingFetched.current = newPages
            },
            clearFetchedPage: () => {
                pagesFetched.current = []
            },
        }),
        [],
    )
}

async function attemptTo(fn: Function) {
    let failed = false
    let result
    let error
    try {
        result = await fn()
    } catch (err) {
        failed = true
        error = err
    }
    return {failed, error, result}
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noop() {}
