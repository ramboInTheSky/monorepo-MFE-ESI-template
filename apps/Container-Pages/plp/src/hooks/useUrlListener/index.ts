import {useEffect} from "react"
import {HISTORY_EVENT} from "../../config/constants"
import {getWindow} from "../../utils/window"

export const useUrlListener = (loadPageFromUrl: (url: string) => void) => {
    useEffect(() => {
        getWindow()!.history.replaceState({type: HISTORY_EVENT}, "", getWindow()!.location.toString())

        const handleUrlChange = (event: PopStateEvent) => {
            if (event.state && event.state.type === HISTORY_EVENT) {
                loadPageFromUrl(getWindow()!.location.toString())
            } else {
                getWindow()!.history.go(-1)
            }
        }
        getWindow()!.addEventListener("popstate", handleUrlChange)
        return () => {
            getWindow()!.removeEventListener("popstate", handleUrlChange)
        }
    }, [loadPageFromUrl])
}
