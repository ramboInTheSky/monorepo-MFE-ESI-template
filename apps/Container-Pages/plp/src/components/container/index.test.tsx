import React from "react"
import {render, cleanup} from "@testing-library/react"
import {ThemeProvider} from "styled-components"
import {mockTheme} from "../../../__mocks__/mockStore"
import {PlpContainer} from "."
import {useUrlListener} from "../../hooks/useUrlListener"

jest.mock("../../hooks/useUrlListener", () => ({
    useUrlListener: jest.fn(),
}))

const mockLoadPageFn = jest.fn()

describe("Container: ", () => {
    afterEach(() => {
        cleanup()
    })

    it("should match the snapshot ", () => {
        const {asFragment} = render(
            <ThemeProvider theme={mockTheme}>
                <PlpContainer loadPageFromUrl={mockLoadPageFn}>
                    <div>Test</div>
                </PlpContainer>
                ,
            </ThemeProvider>,
        )
        expect(asFragment()).toMatchSnapshot()
    })
    it("should call url listener hook", () => {
        expect(useUrlListener).toHaveBeenCalledWith(mockLoadPageFn)
    })
})
