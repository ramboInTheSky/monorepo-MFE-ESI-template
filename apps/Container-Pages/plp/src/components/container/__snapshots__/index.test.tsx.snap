// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container:  should match the snapshot  1`] = `
<DocumentFragment>
  .c0.c0 {
  padding: 0;
}

<div
    class="MuiContainer-root c0 MuiContainer-maxWidthXl"
    id="plp"
  >
    <div>
      Test
    </div>
  </div>
  ,
</DocumentFragment>
`;
